### PR TITLE
Update grammar `functionDefinition` parser rule and rename `NumberUnit` lexer rule

### PR DESCRIPTION
**Pull request by matheusaaguiar**
While working in [literal suffixes](https://github.com/ethereum/solidity/pull/14054), some problematic cases not covered by tests were found and [motivated](https://github.com/ethereum/solidity/pull/14054#issuecomment-1477958615) this.

### CODE REVIEW

File: docs/grammar/SolidityLexer.g4 [hunk 1]
```diff
@@ -62,7 +62,7 @@ New: 'new';
 /**
  * Unit denomination for numbers.
  */
-NumberUnit: 'wei' | 'gwei' | 'ether' | 'seconds' | 'minutes' | 'hours' | 'days' | 'weeks' | 'years';
+Subdenomination: 'wei' | 'gwei' | 'ether' | 'seconds' | 'minutes' | 'hours' | 'days' | 'weeks' | 'years';
```
Comment (nikola-matic) [inline docs/grammar/SolidityLexer.g4]:
```suggestion
SubDenomination: 'wei' | 'gwei' | 'ether' | 'seconds' | 'minutes' | 'hours' | 'days' | 'weeks' | 'years';
```

File: docs/grammar/SolidityLexer.g4 [hunk 2]
```diff
@@ -220,6 +220,12 @@ fragment EvenHexDigits: HexCharacter HexCharacter ('_'? HexCharacter HexCharacte
 //@doc:inline
 fragment HexCharacter: [0-9A-Fa-f];
 
+/**
+ * Scanned but not used by any rule, i.e, disallowed.
+ */
+OctalNumber: '0' DecimalDigits ('.' DecimalDigits)?;
```
Comment (nikola-matic) [inline docs/grammar/SolidityLexer.g4:228]: Technically, `8` and `9` are not valid octal numbers, but are present in `DecimalDigits` (at least I'd hope, I haven't looked :D). So something like `091` would not be a valid octal number in the first place, but the parser error would indicate that octal numbers are not supported, which could be a bit confusing if someone makes a typo, e.g. `091` instead of `91`. I think we might have to introduce `OctalDigits` as well if we want to be fully correct, at least I'm assuming that's something @ekpyron or @cameel would say.

File: docs/grammar/SolidityParser.g4 [hunk 3]
```diff
@@ -175,6 +175,17 @@ locals[
 	 )*
 	(Returns LParen returnParameters=parameterList RParen)?
 	(Semicolon | body=block);
+
+/** The definition of a free function.
+ *
+ */
```
Comment (nikola-matic) [inline docs/grammar/SolidityParser.g4]:
```suggestion
/**
 * The definition of a free function.
 */
```

File: docs/grammar/SolidityParser.g4 [hunk 4]
```diff
@@ -175,6 +175,17 @@ locals[
...
+ freeFunctionDefinition:
```
Comment (nikola-matic) [inline docs/grammar/SolidityParser.g4:182]: I feel like this and the `NumberUnit` changes should be separate PRs, since it's a bit confusing this way, especially since the tests only cover the `NumberUnit` changes.

Comment (cameel) [reply docs/grammar/SolidityParser.g4:182]: Yeah, would not mind having those as separate PRs, though I'm also fine with one because we already have quite a few suffix-related PRs :)

File: docs/grammar/SolidityLexer.g4 (hunk 2 above)
Comment (cameel) [reply docs/grammar/SolidityLexer.g4:228]: I think we can leave it as is because solc's parser behaves the same way. I.e. `091` results in
```
Error: Octal numbers not allowed.
```

Perhaps we should change that in solc, but that's a separate matter - and it would still be an error, just a different one (what is even `091`?).

In any case, we should definitely have a test case for something like `091` to make sure both the grammar and the parser at least both treat it as an error.

Comment (nikola-matic) [reply docs/grammar/SolidityLexer.g4:228]: Fair enough. But I can tell you what `091` *isn't*. An octal number :D

Comment (cameel) [reply docs/grammar/SolidityLexer.g4:228]: We could at least add a note about this weirdness in the docstring above.

Comment (matheusaaguiar) [reply docs/grammar/SolidityLexer.g4:228]: I was mainly focused in mirroring `solc` parser, which seems to treat anything starting with a `0` (and not followed by `.`) as octal...
I am going to add the test case regarding `091`.

File: docs/grammar/SolidityParser.g4 (hunk 4 above)
Comment (matheusaaguiar) [reply docs/grammar/SolidityParser.g4:182]: I felt that, since we already tested (and cameel reviewed) the changes in the literal suffix PR, it would be easier to just push this as a single one. Plus, not too many changes.
But, no problem splitting it, if we consider that should be done.

File: docs/grammar/SolidityLexer.g4 (hunk 2 above)
Comment (matheusaaguiar) [reply docs/grammar/SolidityLexer.g4:228]: Also added the note in the docstring.

File: docs/grammar/SolidityParser.g4 [hunk 5]
```diff
@@ -175,6 +175,17 @@ locals[
 	 )*
 	(Returns LParen returnParameters=parameterList RParen)?
 	(Semicolon | body=block);
+
+/**
+ * The definition of a free function.
+ */
+ freeFunctionDefinition:
+ 	Function identifier
```
Comment (cameel) [inline docs/grammar/SolidityParser.g4]: `fallback` and `receive` seem to be allowed for free functions (they only generate a warning).

```suggestion
 	Function (identifier | Fallback | Receive)
```

So we probably don't have any test cases covering this? Please add some if we don't.

File: test/libsolidity/syntaxTests/literals/invalid_octal_string_no_whitespace.sol [hunk 6]
```diff
@@ -0,0 +1,5 @@
+contract C {
+    uint y = 01abc;
```
Comment (cameel) [inline test/libsolidity/syntaxTests/literals/invalid_octal_string_no_whitespace.sol]: I'd use a denomination here so that the only thing wrong in the test is the thing we're testing for:

```suggestion
    uint y = 01gwei;
```

**Review by cameel — COMMENTED** [review]
Needs only small corrections.

File: docs/grammar/SolidityParser.g4 (hunk 5 above)
Comment (matheusaaguiar) [reply docs/grammar/SolidityParser.g4]: Oh, I just assumed that they would not be allowed. I am going to add the tests.

File: test/libsolidity/syntaxTests/literals/invalid_octal_string_no_whitespace.sol (hunk 6 above)
Comment (matheusaaguiar) [reply test/libsolidity/syntaxTests/literals/invalid_octal_string_no_whitespace.sol]: Ok.I will then also rename the file to `invalid_octal_denomination_no_whitespace.sol`.

**Review by cameel — DISMISSED** [review]
Looks good. Please squash review fixes into the original commits and we can merge.

Also, you should add a changelog entry for some of the grammar changes. The renamings and splits are just refactors but we also have some actual changes:
- a fix of the discrepancy between the grammar and our parser for numbers followed by an identifier
- grammar is now more strict for free function definitions and will no longer accept some things that are only allowed for contract functions

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `9343b81f6b1e17c088f2f420c7a8afaaa0c236be`:
```
Error: Trailing whitespace found:
Changelog.md:16: 

```
Please check that your changes are working as intended.

File: Changelog.md [hunk 7]
```diff
@@ -10,7 +10,10 @@ Compiler Features:
 
 
 Bugfixes:
-
```
Comment (matheusaaguiar) [inline Changelog.md]: Not sure if it should have its own section like:
```
Antlr Grammar Changes:
 * Lexer: Fix of a discrepancy which allowed numbers followed by an identifier with no whitespace
 * Lexer: Fix divergent behaviour which allowed octal numbers.
 * Parser: More strict for free functions and will no longer accept some specifiers which are exclusive of contract functions.
```

**Comment by matheusaaguiar** [comment]
I was not sure if the changes should go in their own section titled `Antlr Changes`.

File: Changelog.md [hunk 8]
```diff
@@ -10,6 +10,9 @@ Compiler Features:
 
 
 Bugfixes:
+ * Antlr Grammar: Fix of a discrepancy which allowed numbers followed by an identifier with no whitespace.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Antlr Grammar: Fix of a discrepancy with the parser, which allowed numbers followed by an identifier with no whitespace.
```

File: Changelog.md [hunk 9]
```diff
@@ -10,6 +10,9 @@ Compiler Features:
...
+ * Antlr Grammar: Stricter rules for function definitions will no longer accept as valid free functions having some specifiers which are exclusive of contract functions.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Antlr Grammar: Stricter rules for function definitions. The grammar will no longer accept as valid free functions having specifiers which are exclusive to contract functions.
```

File: Changelog.md [hunk 10]
```diff
@@ -10,6 +10,9 @@ Compiler Features:
...
+ * Antlr Grammar: Fix divergent behaviour which allowed octal numbers.
```
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Antlr Grammar: Fix discrepancy with the parser, which allowed octal numbers.
```

**Review by cameel — COMMENTED** [review]
Some text corrections.

File: Changelog.md (hunk 7 above)
Comment (cameel) [reply Changelog.md]: Let's follow the way it was done in older entries.

File: Changelog.md (hunk 10 above)
Comment (nikola-matic) [reply Changelog.md]: This should go one up in order to keep the changelog sorted.

**Comment by cameel** [comment]
By the way, just FYI, you don't have to squash it into a single commit. In this case it would be even nicer to have the independent changes in separate commits. There should just be no commits that change things from other commits from the same PR. Other than that it mostly matters that each commit is a single logical change.

A good rule is that if it's hard to describe the set of changes in one short sentence, it's probably doing more than one thing and could be split.

**Comment by matheusaaguiar** [comment]
> By the way, just FYI, you don't have to squash it into a single commit. In this case it would be even nicer to have the independent changes in separate commits. 

Good to know that. I'll keep in mind for the future. Thanks.
Should we leave this in a single commit or try to restore the order from before?

**Review by cameel — APPROVED** [review]

**Comment by cameel** [comment]
Let's leave it as is so that we can merge it :)
